Worklogs CRUD feature

prabeg-introcept opened the pull request:
- Create: Add migrations for worklogs and set 1:n relationship between User and Worklog
- Create: Add validation for worklog creation
- Create: Add view for create worklog
- Create and Read: Store worklog, redirect user to dashboard with success message
- Read: Add view for individual worklog
- Update: Add request validation for edit worklog form
- Udpate: Add action to update worklogs

prabeg-introcept committed:
Worklogs CRUD feature

- Create: Add migrations for worklogs and set 1:n relationship between User and Worklog
- Create: Add validation for worklog creation
- Create: Add view for create worklog
- Create and Read: Store worklog, redirect user to dashboard with success message
- Read: Add view for individual worklog
- Update: Add request validation for edit worklog form
- Udpate: Add action to update worklogs

ashish-shakya commented on app/Http/Controllers/Users/WorklogController.php:
```diff
@@ -17,72 +21,74 @@ class WorklogController extends Controller
     {
-        //
-    }
+        Worklog::create($request->validated());
```
better to wrap this action with try catch exception handler as well

chandan-introcept commented on database/migrations/2021_11_16_104034_create_worklogs_table.php:
```diff
@@ -0,0 +1,38 @@
+            $table->text('description')->nullable();
+            $table->foreignId('user_id')
+                ->constrained(DbTables::USERS)
+                ->onDelete('cascade')
+                ->onUpdate('cascade');
```
we can use cascadeOnDelete() built int method

ashish-shakya commented on app/Http/Controllers/Users/WorklogController.php:
```diff
@@ -17,72 +21,74 @@ class WorklogController extends Controller
+
+        $updatedWorklog = $request->validated();
+
+        $worklog->title = $updatedWorklog['title'];
```
it can be simplified as `$worklog->update($arrayOfNewDataSet)` or simply `$worklog->update($request->validated())`

chandan-introcept commented on app/Http/Controllers/Users/WorklogController.php:
```diff
@@ -17,72 +21,74 @@ class WorklogController extends Controller
     {
-        //
-    }
+        $dateCreated = strtotime(date('Y-m-d', strtotime($worklog->created_at)));
+        $dateToday = strtotime(date('Y-m-d'));
```
we can use Carbon package for manupulating dates

ashish-shakya commented on app/Http/Requests/StoreWorklogRequest.php:
```diff
@@ -0,0 +1,34 @@
+    {
+        return [
+            'title' => ['required', 'string', 'max:80'],
+            'description' => ['nullable', 'string'],
```
better to define max validation rule as well, Since description would be filled in text area, its better to define character limit as well

ashish-shakya commented on app/Http/Requests/UpdateWorklogRequest.php:
```diff
@@ -0,0 +1,34 @@
+    {
+        return [
+            'title' => ['required', 'string', 'max:80'],
+            'description' => ['nullable', 'string'],
```
same goes here. give max lenght validatio in description field

chandan-introcept commented on app/Http/Controllers/Users/WorklogController.php:
```diff
@@ -17,72 +21,74 @@ class WorklogController extends Controller
+
+        $updatedWorklog = $request->validated();
+
+        $worklog->title = $updatedWorklog['title'];
+        $worklog->description = $updatedWorklog['description'];
+
+        $worklog->save();
```
$worklog->update($updateWorklog)

ashish-shakya commented on database/migrations/2021_11_16_104034_create_worklogs_table.php:
```diff
@@ -0,0 +1,38 @@
+    {
+        Schema::create('worklogs', function (Blueprint $table) {
+            $table->id();
+            $table->string('title', 80);
```
since u have defined the char limit for title, make sure u define same validation rule in request class as well

chandan-introcept commented on app/Http/Requests/StoreWorklogRequest.php:
```diff
@@ -0,0 +1,34 @@
+     *
+     * @return bool
+     */
+    public function authorize()
+    {
+        return Auth::check();
+    }
+
```
I think need to check Auth::check() simply return true

ashish-shakya commented on database/migrations/2021_11_16_104034_create_worklogs_table.php:
```diff
@@ -0,0 +1,38 @@
+     */
+    public function up()
+    {
+        Schema::create('worklogs', function (Blueprint $table) {
```
use DBTables constanst for db name as per the previous convention

ashish-shakya commented on database/migrations/2021_11_16_104034_create_worklogs_table.php:
```diff
@@ -0,0 +1,38 @@
+     */
+    public function down()
+    {
+        Schema::dropIfExists('worklogs');
```
use DBTables constanst for db name as per the previous convention

ashish-shakya commented on resources/views/templates/flash-message.blade.php:
```diff
@@ -0,0 +1,30 @@
+
+@if ($message = Session::get('error'))
+    <div class="alert alert-danger alert-dismissible fade show" role="alert">
+        <strong>{{ $message }}</strong>
```
can we refactor this file, I see only class is different based on session key, is there any way we can resuse the other piece of code.

ashish-shakya replied:
may be we can define in some array and match the class based on session key and the set it dynamically in class attribute

chandan-introcept commented on resources/views/user/dashboard.blade.php:
```diff
@@ -1,6 +1,36 @@
+    </tr>
+    </thead>
+    <tbody>
+
+    @foreach($worklogs as $worklog)
+    <tr>
+        <td>
+            <?= date('Y-m-d, h:i A', strtotime($worklog->created_at))?>
```
for manupulating date we can use carbon or simply $worklog->created_at->format('Y-m-d, h:i A')

ashish-shakya replied:
cant we replace this core style with blade templating?

chandan-introcept commented on resources/views/worklog/edit.blade.php:
```diff
@@ -0,0 +1,59 @@
+            type="hidden"
+            class="form-control"
+            name="user_id"
+            value="{{\Illuminate\Support\Facades\Auth::id()}}"
```
auth()->id()

prabeg-introcept committed:
Changes for PR review

- Implement service class for worklogs and handle exceptions in controller
- Change the description column to tinyText and add validation for it in form request
- Replace PHP with Blade syntax
- Implement DRY for flash messages section
- Add constant for worklogs to be used in database files

chandan-introcept commented on app/Services/WorklogService.php:
```diff
@@ -0,0 +1,73 @@
+        $worklogs = Worklog::where('user_id', Auth::id())
+            ->orderBy('created_at', 'desc')
+            ->get();
+        if(!$worklogs)
+        {
+            throw new WorklogNotFoundException('No worklogs created yet. Click Add Task to create new worklog');
+        }
```
instead of throwing exception. we can show message "No worklogs created yet. Click Add Task to create new worklog" on ui

prabeg-introcept replied:
Yes, the exception message thrown from here is being rendered in the UI via the controller within the catch block.

chandan-introcept commented on app/Services/WorklogService.php:
```diff
@@ -0,0 +1,73 @@
+    {
+        $worklog = $this->getWorklog($worklogId);
+
+        $dateCreated = Carbon::now()->format('Y-m-d');
+        $dateToday = $worklog->created_at->format('Y-m-d');
+
+        if($dateCreated !== $dateToday){
```
no need of formatting ,we can check like  ` $worklog->created_at->isToday() `

chandan-introcept commented on resources/views/user/dashboard.blade.php:
```diff
@@ -17,7 +17,7 @@
     <tr>
         <td>
-            <?= date('Y-m-d, h:i A', strtotime($worklog->created_at))?>
+            {{ date('Y-m-d, h:i A', strtotime($worklog->created_at)) }}
```
$worklog->created_at->format('Y-m-d, h:i A')

chandan-introcept commented on app/Http/Controllers/Users/WorklogController.php:
```diff
@@ -46,21 +62,31 @@ public function create()
+    public function edit($id)
     {
+        try{
+            $worklog = $this->worklogService->getWorklog($id);
```
$this->worklogService->getWorklog($id); -> this one is already throwing not found exception

prabeg-introcept replied:
here, the model not found exception thrown redirects to Laravel custom 404 page, but to show that worklog with the ID does not exist I have used the try-catch block to show the error message in the same view rather than Laravel custom 404 page.

chandan-introcept commented on app/Services/WorklogService.php:
```diff
@@ -0,0 +1,73 @@
+     */
+    public function getLoggedInUserWorklogs(): Collection
+    {
+        $worklogs = Worklog::where('user_id', Auth::id())
+            ->orderBy('created_at', 'desc')
+            ->get();
```
`auth()->user()->workflows`

prabeg-introcept committed:
Remove duplication of onUpdateCascade, and add onDeleteCascade

prabeg-introcept committed:
Fetch worklogs for a user in descending order of created date

prabeg-introcept committed:
Update flash messages

prabeg-introcept committed:
Inject worklog dependancy, use the throw_if helper, and clean methods name

prabeg-introcept committed:
Make changes according to WorklogService

prabeg-introcept committed:
Simplify date formatting
